ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/parallelization/BalConcurrentTransactionProcessor.java

```diff
@@ -0,0 +1,227 @@
+      ctxBuilder.transactionAccumulator(blockUpdater).transactionProcessingResult(result);
+
+      return ctxBuilder.build();
+    } finally {
+      ws.close();
+    }
+  }
+
+  @Override
+  // TODO: Throw instead of returning Optional.empty()?
+  public Optional<TransactionProcessingResult> getProcessingResult(
+      final MutableWorldState worldState,
+      final Address miningBeneficiary,
+      final Transaction transaction,
+      final int txIndex,
+      final Optional<Counter> confirmedParallelizedTransactionCounter,
+      final Optional<Counter> conflictingButCachedTransactionCounter) {
+
+    final CompletableFuture<ParallelizedTransactionContext> future = futures[txIndex];
+    if (future != null) {
+      try {
+        final ParallelizedTransactionContext ctx =
+            balProcessingTimeout.isNegative()
+                ? future.join()
+                : future.get(balProcessingTimeout.toNanos(), TimeUnit.NANOSECONDS);
+
+        if (ctx == null) {
+          LOG.error("Transaction context for transaction {} is empty.", txIndex);
+          return Optional.empty();
+        }
+
+        final PathBasedWorldState pathWs = (PathBasedWorldState) worldState;
+        final PathBasedWorldStateUpdateAccumulator blockAccumulator =
+            (PathBasedWorldStateUpdateAccumulator) pathWs.updater();
+
+        final PathBasedWorldStateUpdateAccumulator<?> txAccumulator = ctx.transactionAccumulator();
+        final TransactionProcessingResult result = ctx.transactionProcessingResult();
+
+        blockAccumulator.importStateChangesFromSource(txAccumulator);
+
+        confirmedParallelizedTransactionCounter.ifPresent(Counter::inc);
+        result.setIsProcessedInParallel(Optional.of(Boolean.TRUE));
+        result.accumulator = txAccumulator;
+
+        return Optional.of(result);
+      } catch (final TimeoutException e) {
+        LOG.error(
+            "Timed out waiting {}ms for transaction {} processing result.",
+            balProcessingTimeout.toMillis(),
+            txIndex);
+        return Optional.empty();
+      } catch (final Exception e) {
+        LOG.error(
+            "Error integrating transaction processing result for transaction {}.", txIndex, e);
+        return Optional.empty();
+      }
+    }
+
+    LOG.error("No future found for transaction {}.", txIndex);
+    return Optional.empty();
+  }
+
+  private void applyWritesFromPriorTransactions(
+      final BlockAccessList blockAccessList,
+      final int balIndex,
+      final PathBasedWorldStateUpdateAccumulator<?> worldStateUpdater) {
+
+    blockAccessList
+        .accountChanges()
+        .forEach(
+            accountChanges -> {
+              final Address address = accountChanges.address();
+
+              accountChanges.balanceChanges().stream()
+                  .filter(change -> change.txIndex() < balIndex)
+                  .max(Comparator.comparingInt(BlockAccessList.BalanceChange::txIndex))
```

Comment:
We can make all the comparators static, ex.
private static final Comparator<BlockAccessList.BalanceChange> BALANCE_CHANGE_COMPARATOR =
    Comparator.comparingInt(BlockAccessList.BalanceChange::txIndex);